### PR TITLE
[v0.9] Reduce BundleDeployment triggering on deployed resources updates

### PR DESCRIPTION
**Pull request by aruiz14**
Refers to #2058

These changes aim for reducing the number of time that the BundleDeployment reconciler is executed. This happens due to the Fleet Agent configuring a watcher on every deployed resources, which will trigger a requeue of the BundleDeployment handler in the agent for every watch event. Since this handler refreshes the conditions timestamp in the BundleDeployment status, this has a direct effect on the Fleet controller as well, which will react to the update in the BundleDeployment object.

This patch consists of the following changes:
 - The `TriggerSleep` delay has been removed, since it could cause modification events happening in that time window to be missed (this was the case for the integration tests). This is connected to the next change:
 - `ADDED` watch events are now ignored:
   - The informer creates an initial `ADDED` event for every object, but this does not make sense for objects that were just created. 
   - The only interesting events are `MODIFIED` and `DELETED`, which are the ultimate interest of the trigger (detecting modified or deleted resources). If an object is deleted and later created by the user, the controller may still need to redeploy in order to achieve a stable situation.
- Filters modified events based on `Generation` field, for those kinds that make use of it.
  - Kinds that do not implement this metadata field (e.g. ConfigMaps) will keep the existing behavior.
  - Resources that set the `Generation` field will be checked: based on their UID, if `Generation` didn't change from a previous event, it will be interpreted as the spec didn't change, so will skip the trigger, effectively omitting status updates.

### CODE REVIEW

**Comment by raulcabello** [comment]
I like the approach. This introduces two improvements:
- Removing `ADDED` -> removes an initial enqueue that happens for all resources being watched as the informer creates an initial ADDED event for each resource.
- Filters modified events based on Generation field -> removes a lot of enqueues that happens when status fields are changed. We ignore status changes when detecting drift in [wrangler](https://github.com/rancher/wrangler/blob/master/pkg/apply/desiredset_compare.go#L135-L138)

I don't see any issue with this approach. However, there is a possibility that these extra enqueues are needed in some edge cases we don't see at the moment. So, I wouldn't add this to the stable release (0.8)

File: internal/cmd/agent/trigger/watcher.go [hunk 1]
```diff
@@ -125,10 +131,36 @@ func (t *Trigger) OnChange(key string, defaultNamespace string, trigger func(),
 	return nil
 }
 
-func (t *Trigger) call(gvk schema.GroupVersionKind, key objectset.ObjectKey) {
+func (t *Trigger) call(gvk schema.GroupVersionKind, obj metav1.Object, deleted bool) {
```
Comment (weyfonk) [inline internal/cmd/agent/trigger/watcher.go:134]: How feasible would it be to cover this with tests? I for one would then feel more comfortable that this reduces triggering without introducing any important, undesired side-effects.
eg.:
* deleted object → triggers called
* added object → triggers not called
* modified object with no generation field → triggers called
* modified object with generation field set to 0 → triggers not called
* modified object with generation field updated → triggers called (possibly repeated a couple of times to validate that subsequent generation updates have the desired effect)

Comment (aruiz14) [reply internal/cmd/agent/trigger/watcher.go:142]: Tests added, thanks for the suggestion. It made me find that there was a corner case in which a trigger would happen but is not desired. Please take another look!

Comment (aruiz14) [reply internal/cmd/agent/trigger/watcher.go:142]: One of the tests is flaky, I'm looking into it

File: integrationtests/agent/watcher_trigger_test.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+		It("is not initially triggered", func() {
+			Consistently(func() int {
+				return *triggerCount
+			}).Should(Equal(0))
+		})
+		It("is triggered on deletion", func() {
+			Expect(env.k8sClient.Delete(ctx, cm)).ToNot(HaveOccurred())
+			Eventually(func() int {
+				return *triggerCount
+			}).WithPolling(100 * time.Millisecond).MustPassRepeatedly(3).
```
Comment (weyfonk) [inline integrationtests/agent/watcher_trigger_test.go:79]: Out of curiosity, where does this `3` come from?

File: integrationtests/agent/watcher_trigger_test.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+					Message:            "tests",
+					Reason:             "tests",
+				},
+			}
+			Expect(env.k8sClient.Status().Update(ctx, deploy)).ToNot(HaveOccurred())
+			Consistently(func() int {
+				return *triggerCount
+			}).Should(Equal(0))
+		})
+		It("is triggered Spec is changed", func() {
```
Comment (weyfonk) [inline integrationtests/agent/watcher_trigger_test.go:182]: nit:
```suggestion
		It("is triggered when Spec is changed", func() {
```
or (more consistent with the previous test description):
```suggestion
		It("is triggered on Spec changes", func() {
```

**Review by weyfonk — APPROVED** [review]
LGTM, thanks for taking the time to add integration tests for this :relaxed:

File: integrationtests/agent/watcher_trigger_test.go (hunk 2 above)
Comment (aruiz14) [reply integrationtests/agent/watcher_trigger_test.go:79]: It's a way to express that the trigger count should eventually increase from 0 -> 1, and not increase any longer. In other words, just executed once.
The actual 3 and the polling interval are totally made up values, I didn't want to slow down the tests too much either.
Happy to apply any suggestions you may have!

File: internal/cmd/agent/trigger/watcher.go [hunk 4]
```diff
@@ -125,10 +132,42 @@ func (t *Trigger) OnChange(key string, defaultNamespace string, trigger func(),
...
+			t.seenGenerations.Delete(uid)
+		} else {
+			// keep a map of UID -> generation, using sync.Map and atomic.Int64 for safe concurrent usage
+			// - sync.Map entries are never modified after created, a pointer is used as value
+			// - using atomic.Int64 as values allows safely comparing and updating the current Generation value
+			var previous *atomic.Int64
+			if value, ok := t.seenGenerations.Load(uid); ok {
+				previous = value.(*atomic.Int64)
+			} else {
+				previous = t.storeObjectGeneration(uid, currentGeneration)
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply internal/cmd/agent/trigger/watcher.go:157]: Hmm it makes sense, let me rework it.

File: internal/cmd/agent/trigger/watcher.go (hunk 1 above)
Comment (aruiz14) [reply internal/cmd/agent/trigger/watcher.go:142]: Fixed in ad49a1638a0aa564300d3cd290e0dbf97da64e3c

File: integrationtests/agent/watcher_trigger_test.go (hunk 2 above)
Comment (weyfonk) [reply integrationtests/agent/watcher_trigger_test.go:78]: No worries, I was just wondering how you came to `3` instead of, say, 5 or 10 (for instance, also made up values).
3 works for me :)

File: internal/cmd/agent/trigger/watcher.go [hunk 5]
```diff
@@ -125,10 +132,40 @@ func (t *Trigger) OnChange(key string, defaultNamespace string, trigger func(),
 	return nil
 }
 
-func (t *Trigger) call(gvk schema.GroupVersionKind, key objectset.ObjectKey) {
+func (t *Trigger) storeObjectGeneration(uid types.UID, generation int64) *atomic.Int64 {
+	value := new(atomic.Int64)
+	value.Store(generation)
+	t.seenGenerations.Store(uid, value)
+	return value
+}
```
Comment (weyfonk) [inline internal/cmd/agent/trigger/watcher.go:140]: I think this would make the linter happy:
```suggestion
func (t *Trigger) storeObjectGeneration(uid types.UID, generation int64) {
	value := new(atomic.Int64)
	value.Store(generation)
	t.seenGenerations.Store(uid, value)
}
```
